domain/schema/model/sql/0020-network.sql

```diff
@@ -14,9 +14,295 @@ CREATE UNIQUE INDEX idx_cloud_service_application
+CREATE TABLE ip_address_config_type (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_ip_address_config_type_name
+ON ip_address_config_type (name);
+
+INSERT INTO ip_address_config_type VALUES
+(0, 'unknown'),
+(1, 'dhcp'),
+(2, 'static'),
+(3, 'manual'),
+(4, 'loopback');
+
+CREATE TABLE ip_address (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    -- The value of the configured IP address.
+    -- e.g. 192.168.1.2 or 2001:db8::1.
+    address_value TEXT NOT NULL,
+    -- one of ipv4, ipv6 etc.
+    type_id INT NOT NULL,
+    -- one of dhcp, static, manual, loopback etc.
+    config_type_id INT NOT NULL,
+    -- one of host, provider
+    origin_id INT NOT NULL,
+    -- one of public, local-cloud, local-machine, link-local etc.
+    scope_id INT NOT NULL,
+    -- the link layer device this address belongs to.
+    device_uuid TEXT NOT NULL,
+
+    -- indicates that this address is not the primary.
+    -- address associated with the NIC.
+    is_secondary BOOLEAN DEFAULT false,
+
+    -- indicates whether this address is a virtual/floating/shadow.
+    -- address assigned to a NIC by a provider rather than being
+    -- associated directly with a device on-machine.
+    is_shadow BOOLEAN DEFAULT false,
+
+    CONSTRAINT fk_ip_address_link_layer_device
+    FOREIGN KEY (device_uuid)
+    REFERENCES link_layer_device (uuid),
+    CONSTRAINT fk_ip_address_origin
+    FOREIGN KEY (origin_id)
+    REFERENCES ip_address_origin (id),
+    CONSTRAINT fk_ip_address_type
+    FOREIGN KEY (type_id)
+    REFERENCES ip_address_type (id),
+    CONSTRAINT fk_ip_address_config_type
+    FOREIGN KEY (config_type_id)
+    REFERENCES ip_address_config_type (id),
+    CONSTRAINT fk_ip_address_scope
+    FOREIGN KEY (scope_id)
+    REFERENCES network_address_scope (id)
+);
+
+CREATE TABLE ip_address_dns_server_address (
+    address_uuid TEXT NOT NULL,
+    dns_server_address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_ip_address_dns_server_address_dns_server_address
+    FOREIGN KEY (dns_server_address_uuid)
+    REFERENCES ip_address (uuid),
+    CONSTRAINT fk_ip_address_dns_server_address_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid),
+    PRIMARY KEY (dns_server_address_uuid, address_uuid)
+);
+
+CREATE TABLE ip_address_dns_search_domain (
+    address_uuid TEXT NOT NULL,
+    search_domain TEXT NOT NULL,
+    CONSTRAINT fk_ip_address_dns_search_domain_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid),
+    PRIMARY KEY (address_uuid, search_domain)
+);
+
+CREATE TABLE fqdn_address (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    hostname TEXT NOT NULL,
+    -- one of public, local-cloud, local-machine, link-local etc.
+    scope_id INT NOT NULL,
+
+    CONSTRAINT fk_fqdn_address_scope
+    FOREIGN KEY (scope_id)
+    REFERENCES network_address_scope (id)
+);
+
+CREATE UNIQUE INDEX idx_fqdn_address_hostname
+ON fqdn_address (hostname);
+
+CREATE TABLE net_node_fqdn_address (
```

Comment:
do you not get the `net_node_uuid` relation via the address having a device? Seems like we have two paths for the same link that could be in conflict?

Reply:
I think seeing how this is supposed to work in reality and the relation ships that we have we want to change all the wording around fqdn to just hostname?

Hostname is discovered and forms the relationship. FQDN is something else created outside of Juju's purview and there is no guarantee unless we run an eventually consistent resolver that a fqdn will always resolve to this information.

Reply:
We really need 2 tables - one for hostname and one for fqdn.